### PR TITLE
Ppinheiro/log support

### PR DESCRIPTION
**Pull request by paulovap**
Save qDebug and friends into a log file

### CODE REVIEW

File: gui/src/logs.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    // function, so it can find the right app folder.
+    log_path =
+        QStandardPaths::writableLocation(QStandardPaths::AppLocalDataLocation);
+#ifdef QT_DEBUG
+    printf("saving logs in %s\n", log_path.toLatin1().data());
+#endif
+    printf("log folder created? %d\n", QDir().mkpath(log_path));
+  }
+  QFile outFile(log_path + "/gui.log");
+  outFile.open(QIODevice::WriteOnly | QIODevice::Append);
```
Comment (martukas) [inline gui/src/logs.h:25]: So, I assume this handler gets called every time a log message is written, and then it opens the file anew every time? Is there not a way to open this sink once and have it persist?

File: gui/src/logs.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#include <QTextStream>
+
+static QString log_path = "";
+
+void logOutputToFile(QtMsgType type, const QMessageLogContext &context,
+                     const QString &msg) {
+  if (log_path.isEmpty()) {
+    // log_path assignment needs to be delayed until first usage of this
+    // function, so it can find the right app folder.
+    log_path =
```
Comment (martukas) [inline gui/src/logs.h:17]: I vaguely remember the question of path being raised in some channel but then I guess we neglected to have that discussion. Where exactly will this end up? Is it not standard for apps in Linux to keep their logs somewhere under `/var/log` ?

File: gui/src/logs.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    ts << "[WARN]";
+    break;
+  case QtCriticalMsg:
+    ts << "[CRIT]";
+    break;
+  case QtFatalMsg:
+    ts << "[FTAL]";
+    break;
+  }
+  ts << QString(" %1 (%2:%3, %4)")
```
Comment (martukas) [inline gui/src/logs.h:49]: Could we also have a timestamp?

File: gui/src/logs.h (hunk 2 above)
Comment (a-vinod) [reply gui/src/logs.h:17]: Looks like AppLocalDataLocation is evaluated to `"~/.local/share/<APPNAME>", "/usr/local/share/<APPNAME>", "/usr/share/<APPNAME>"` on linux according to the [docs](https://doc.qt.io/qt-5/qstandardpaths.html). Would the GUI have root permissions to write logs into `/var/log`?

File: gui/src/logs.h (hunk 1 above)
Comment (a-vinod) [reply gui/src/logs.h:25]: Could we create and open the log file in the main function before `qInstallMessageHandler` is called? Such as by putting lines 24 and 25 here into the main function and passing `outFile` as a parameter into this function from main?

File: gui/src/logs.h (hunk 2 above)
Comment (martukas) [reply gui/src/logs.h:17]: Ok, I guess this makes sense. No, we shouldn't give it root permissions. Let's keep it there, except maybe add something to a readme about where these logs are to be found.

File: gui/src/logs.h (hunk 1 above)
Comment (martukas) [reply gui/src/logs.h:25]: But main doesn't actually call this function, it only provides this this function as a handler. The function gets called by a macro every time there is something like 
```
qCritical() << "Could not serialize GuiStatus";
```

My instinct is that there should be a singleton that persists with the file open. Main would configure this singleton to open a particular file. The function should then address this singleton.

An example of a good singleton implementation here:
https://github.com/ess-dmsc/daquiri/blob/master/source/core/ProducerFactory.h

However, there is the additional problem that this implementation is not thread-safe. The singleton should probably have a mutex.

**Comment by martukas** [comment]
I'm starting some work on:
* refactoring
* improving
* unit testing
this in:
https://github.com/RespiraWorks/Ventilator/tree/issue_552_gui_logging

**Comment by martukas** [comment]
So, I have reimplemented this using https://github.com/gabime/spdlog
* writes to file
* is thread-safe
* does timestamps
* logging levels are flexible
* write to console when compiled in as debug

As a bonus:
* logs git commit hash upon startup
* `gui.sh` bash script made more flexible, and guards against unwitting runs with sudo
* adds some unit tests for logger (though they kind of suck)
* updated documentation on building and running gui

**Pull request merged**